src/engine/enginemaster.cpp

```diff
@@ -443,48 +530,212 @@ void EngineMaster::process(const int iBufferSize) {
     }
 
     if (masterEnabled) {
-        // Mix the three channels together. We already mixed the busses together
-        // with the channel gains and overall master gain.
-        if (!m_pMasterTalkoverMix->toBool()) {
-            // Add Talkover to Master output
-            SampleUtil::copy4WithGain(m_pMaster,
-                    m_pOutputBusBuffers[EngineChannel::LEFT], 1.0,
-                    m_pOutputBusBuffers[EngineChannel::CENTER], 1.0,
-                    m_pOutputBusBuffers[EngineChannel::RIGHT], 1.0,
+        // Mix the crossfader orientation buffers together into the master mix
+        SampleUtil::copy3WithGain(m_pMaster,
+            m_pOutputBusBuffers[EngineChannel::LEFT], 1.0,
+            m_pOutputBusBuffers[EngineChannel::CENTER], 1.0,
+            m_pOutputBusBuffers[EngineChannel::RIGHT], 1.0,
+            iBufferSize);
+
+        // Process master, booth, and record/broadcast buffers according to the
+        // TalkoverMixMode configured in DlgPrefSound
+
+        // Mixxx gets input signals delayed by the input latency,
+        // so to preserve the relative timing of the input with Mixxx's outputs
+        // (headphones, master, booth, and record/broadcast), delay the outputs
+        // by the input latency before mixing the inputs into them.
+
+        // This way, if the microphone user is on beat with the output they hear from
+        // the speakers/headphones, the input will remain on beat at the expense of
+        // adding some latency to Mixxx's output (no latency will be added to the
+        // microphone signal though). If no microphone inputs are configured, avoid adding
+        // this extra latency.
+
+        // If the input signal is being directly monitored (mixed with the master output
+        // in hardware), do not delay the audible outputs. However, delay the
+        // record/broadcast signal to keep it on beat and consistent with what
+        // the user hears from the speakers/headphones.
+        if (configuredTalkoverMixMode == TalkoverMixMode::MASTER) {
+            // Apply input latency compensation to master mix
+            // Do not add unnecessary latency if no microphones are configured
+            if (m_pNumMicsConfigured->get() > 0) {
+                m_pInputLatencyCompensationDelay->process(m_pMaster, iBufferSize);
+            }
+
+            // Copy master mix to booth output with booth gain before mixing
+            // talkover with master mix
+            if (boothEnabled) {
+                CSAMPLE boothGain = m_pBoothGain->get();
+                if (m_bRampingGain) {
+                    SampleUtil::copy1WithRampingGain(m_pBooth, m_pMaster,
```

Comment:
Which undelayed signal are you referring to? If you are referring to the undelayed input when listening to the delayed booth output, yes, there is nothing to compensate for. But if the latency compensation delay is not processed for the booth, it will be out of time with the master output and create a double beat effect. I experienced this with the PFL signal when working on the original latency compensation implementation in 0f7a50b8441848c13549b724d1a42026dffad21d before creating the EngineDelay for the PFL mix. If any output is delayed, they all must be delayed by the same amount, except for the delays to compensate for speaker placement.